ludwig/data/dataset/tfrecord.py

```diff
@@ -0,0 +1,252 @@
+            if not s3_fs.isdir(url):
+                raise RuntimeError("The remote S3 path should be a folder"
+                                   "containing all tfrecords.")
+            path = tempfile.mkdtemp()
+            for s3_path in s3_fs.ls(url):
+                s3_fs.get(s3_path, path)
+        except ImportError:
+            raise ImportError("Reading from S3 requires `s3fs` support. "
+                              "Please install s3fs following: "
+                              "https://github.com/s3fs-fuse/s3fs-fuse#installation.")
+    elif url.startswith("file"):
+        path = url[7:]
+    else:
+        raise RuntimeError("Unsupported file systems. Support S3 or "
+                           "local FS, but got: `{}`.".format(url))
+    return path
+
+
+class TFRecordDataset(Dataset):
+    def __init__(self, url, features, training_set_metadata):
+        self.url = to_url(url)
+        self.local_url = check_url_and_maybe_download(self.url)
+        self.features = [feature[PROC_COLUMN] for feature in features]
+        self.training_set_metadata = training_set_metadata
+
+        meta = load_json(os.path.join(self.local_url, "meta.json"))
+        self.size = meta["size"]
+        self.compression_type = meta["compression_type"]
+        self.reshape_features = {
+            feature[PROC_COLUMN]: list((-1, *training_set_metadata[feature[NAME]]['reshape']))
+            for feature in features
+            if "reshape" in training_set_metadata[feature[NAME]]
+        }
+
+    def get(self, feature_name, sample):
+        t = sample[feature_name]
+        reshape_dim = self.reshape_features.get(feature_name)
+        if reshape_dim is not None:
+            # When we read a 1D array from disk, we need to reshape it back to its
+            # full dimensions.
+            t = tf.reshape(t, reshape_dim)
+        return t
+
+    def __len__(self):
+        return self.size
+
+    @contextlib.contextmanager
+    def initialize_batcher(self,
+                           batch_size=128,
+                           should_shuffle=True,
+                           seed=0,
+                           ignore_last=False,
+                           horovod=None):
+        cur_shard, shard_count = None, None
+        if horovod:
+            cur_shard, shard_count = horovod.rank(), horovod.size()
+        total_samples = self.size
+        local_samples = int(total_samples / shard_count) if shard_count else total_samples
+
+        # Below are routine optimizations for tf.dataset
+        compression_ext = '.gz' if self.compression_type else ''
+        path = os.path.join(self.local_url, "*.tfrecords{}".format(compression_ext))
+
+        # Now construct the tfrecrd dataset
+        files = tf.data.Dataset.list_files(path)
+
+        # interleave the tfrecord files for parallel reading
+        dataset = files.interleave(
+            lambda x: tf.data.TFRecordDataset(x, compression_type=self.compression_type),
```

Comment:
Looks like this should work with S3 and other remote paths natively: https://blog.min.io/hyper-scale-machine-learning-with-minio-and-tensorflow/